CodeReview of 16 from 8

EliaBedin opened the pull request:
Here's the file for the review.
The introduction header should explain everything, but in case, fell free to PM me.
I wish you good work.

Commit:
Review code

Added the file to be reviewed by Molinaro & Prandini

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+/*
+ *  The main goal for this project is to test older Camera API (15 to 19) to see if it's
+ *  plausible to support these devices and lower the minimum API requirement on CommonDemo
+ *  so we can support full range of Android devices (API 15 and higher is 100% demographic)
```
Do you mean that all the Android devices currently used have an API level equal or higher to 15? Have I understood?

EliaBedin replied:
Yes. When creating a new project in Android Studio you have to set the minimum API requirement for your app; with 15 and higher you are targeting all (it's actually 99.8%, but Android Studio approximates to 100%  and I don't really get why since out of more or less 2 billion Android devices, **2% is still 40000000 devices with API lower than 15**) Android devices.

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+ *  plausible to support these devices and lower the minimum API requirement on CommonDemo
+ *  so we can support full range of Android devices (API 15 and higher is 100% demographic)
+ *  I used this interface for Camera (1) API to make the program
+ *      https://github.com/natario1/CameraView
```
I'm sorry but maybe I've not understood, is this an interface or a library?

EliaBedin replied:
Sorry, it is a bit misleading. It's an Android library that interfaces with Camera API.

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+ *  working with master build of the interface
+ *  The documentation starts here
+ *      https://github.com/natario1/CameraView#usage
+ */
```
I like this introduction because there are the aim of the work and some links about the major tools used.

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+
+        camera.addCameraListener(new CameraListener() {
+            /**
+             * Listining for photo taken by the camera
```
Do you mean "Listening"? Have I understood?

EliaBedin replied:
Yes, it's a typo, thank you for pointing out!

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+     * @author Bedin Elia
+     */
+    private void capturePhoto() {
+        if (mCapturingPicture) return;
```
Should be a good idea to do this disabling the button? Similarly to `buttonPhoto.setEnabled(false);`

EliaBedin replied:
ImageButton does not have a setEnabled, but a android:clickable="false" XML tag as far as I know.
I'll give it a shot soon, thanks you!

PrandiniUniPD replied:
Right, for do this by a xml file but you use an ImageButton object programmatically.
An ImageButton object extends ImageView and ImageView extends View.
In the View object you can enable or disable the state of a View (or a subclassed object, same as ImageButton) programmatically
https://developer.android.com/reference/android/view/View#setEnabled(boolean) 
Could be useful?

EliaBedin replied:
Please forgive my lack of experience with Android.
Your help is really appreciated, I'll read the documentation you suggested me.

PrandiniUniPD replied:
Don't worry I've proposed you only some questions and we've spoken about things, I've to learn with you. So thanks you.

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+     * @author Bedin Elia
+     */
+    @Override
+    public void onRequestPermissionsResult
```
I have not clear where in the code there is the checking of the permissions that eventually launch this method.
Accordingly to this https://developer.android.com/training/permissions/requesting#perm-check it should be similar to this:
`if (ContextCompat.checkSelfPermission(CameraActivity.this, Manifest.permission.CAMERA)`
`        != PackageManager.PERMISSION_GRANTED) {`
`    ActivityCompat.requestPermissions(CameraActivity.this,` 
`                new String[]{Manifest.permission.CAMERA},`
`                MY_PERMISSIONS_REQUEST_CAMERA);`
`}`
Have I understood your intentions? Is it included in the library so it's not necessary?

EliaBedin replied:
The library already do the permission for Camera, but it's implicit like Android K and lower, when there was no promp to ask you.
This code it's for having a visual prompt, and it's an override for the current library method.

PrandiniUniPD replied:
mmm, are you sure? Maybe I'm not understand. Is onRequestPermissionResult for showing a visual prompt for the permissions? https://developer.android.com/reference/android/support/v4/app/ActivityCompat#requestpermissions here it's not seem to be so. In that link I've read that the ActivityCompat.requestPermission launch the visual prompt to the user and the onRequestPermissionsResult catches the results of the request.
But if it is already implemented in the library it's ok.

EliaBedin replied:
Yeah, the library does, see https://github.com/natario1/CameraView#declaration.
Thanks again.

PrandiniUniPD replied:
Ok, thanks to you for this specifications. I've left that point of the documentation! Thanks to you.

EliaBedin replied:
No problem!

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+public class CameraActivity extends AppCompatActivity {
+
+    private CameraView camera;
+    private boolean mCapturingPicture;
```
Is it a good idea to explicitly set this variable to false and put a short description about the future use of this?
But firstly is it a good idea to disable and enable the functionality of the button instead a boolean system?

EliaBedin replied:
You are absolutely right, thank you.
I'll add descriptions and set the variable mCapturingPicture to default 'false'.

> But firstly is it a good idea to disable and enable the functionality of the button instead a boolean system?
See https://github.com/PrandiniUniPD/OCRCamera/pull/64#discussion_r238099864

PrandiniUniPD replied:
Ok, so if you think that it's a good idea enabling and disabling the button and you need some support ask me.

EliaBedin replied:
Thank you very much, I'll keep that in mind!

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+    private void capturePhoto() {
+        if (mCapturingPicture) return;
+        mCapturingPicture = true;
+        camera.capturePicture();
```
I'll try to explain you what I've understood (thanks in advance for the support):
This statement launches the callback for taking a picture and so the method onPictureTaken (line 62) is launched. Finally the onPictureTaken launches the onPicture method (line 95) that firstly sets the boolean variable mCapturingPicture to false. Is it right?
If mCapturingPicture is not already setted to false the capturePhoto method doesn't launch the camera.capturePicture() sentence but returns at the line 85. Is it right?

EliaBedin replied:
You are correct, sorry if it was hard to understand.

PrandiniUniPD replied:
Don't worry, I've asked you for knowing if I've understood the functionality. So thanks for the feedback.

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,123 @@
+                capturePhoto();
+            }
+        });
+
```
Is this empty line permitted here for our Document Style?

EliaBedin replied:
Another typo, thank you for pointing out.

PrandiniUniPD reviewed (commented):
Hi Elia, I've reviewed your code, thanks in advance for your patience, maybe I've not understood everything. So I've commented which are my observations, we could speak about these and finally have a common approved version of your code.

Commit:
Fixed critical aspects pointed in the review

Added variables descriptions.
Changed variable name to make it more intuitive.
Fixed typos and comments

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,126 @@
+     * @author Bedin Elia
+     */
+    private void onCapture(byte[] picture) {
+        isPictureCaptured = false;
```
Could be a good idea making the buttonPhoto a private shared variable for this class, removing isPictureCaptured and the relative statements, putting on this line (99) `buttonPhoto.setEnabled(true);` and putting in the onClick method of the line 77 `buttonPhoto.setEnabled(false)`?

PrandiniUniPD replied:
This is a proposal to avoid a boolean synchronization if it could be useful.
When the button is pressed the button is disabled and it is enabled when the processing is ended.

EliaBedin replied:
Sorry if it took me a long time to apply and test the proposed changes.
They are now applied in the new commit (https://github.com/PrandiniUniPD/OCRCamera/commit/753e923d2dc820a0883776c8fc36e5c847c8a4dc)
Thank you very much for you help on making the code better!

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    /**
+     * Method for sending the taken photo in array form to the PicturePreviewActivity
```
In this and in the next line do you mean form or format?

EliaBedin replied:
I meant "the taken photo in the form of an array (of byte)"

PrandiniUniPD replied:
I've asked you because it seems that the word "form" is mostly used to indicate a shape, not a structure ( https://www.wordreference.com/enit/form ), and I thought you would like to indicate a format instead.

EliaBedin replied:
Fixed in the new pull request.
Thanks for your help!

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,126 @@
+        camera.addCameraListener(new CameraListener() {
+            /**
+             * Listening for photo taken by the camera
+             * @param picture the taken photo in array form (stream of byte)
```
Could you mean format instead of form?

EliaBedin replied:
See https://github.com/PrandiniUniPD/OCRCamera/pull/64#discussion_r239197936

PrandiniUniPD replied:
Other explanations about my question here: https://github.com/PrandiniUniPD/OCRCamera/pull/64#discussion_r239379358

EliaBedin replied:
See https://github.com/PrandiniUniPD/OCRCamera/pull/64#discussion_r239414418

PrandiniUniPD reviewed (commented):
Hi Elia! I've some questions for you, I hope they are useful.

Commit:
Tested changes to the code proposed in the review

Overhaul of the way the app checks if it's ready to take a picture using button.setEnabled().
Changed wording for some comments to make them clearer.

PrandiniUniPD commented on app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java:
```diff
@@ -0,0 +1,132 @@
+             */
+            @Override
+            public void onClick(View view) {
+                capturePhoto();
```
I'm sorry but maybe I wasn't explained well, I've meant that puts between 80 and 81 `buttonPhoto.setEnabled(false);` and removes the line 91-94. In this way the button when pressed once is not being pressed another time until the capturing of the photo is completed and so the capturePhoto method is not launched when the capturing of the photo is running. Could be a good idea?

EliaBedin replied:
In my opinion, it's clearer if the listener don't check anything, just call the method that's supposed to do the action we are listening for.
And to be clear, I'm not saying your proposal is wrong, quite the opposite.
What do you think? Is it so critical that I should move the check inside the listener?

PrandiniUniPD replied:
Rethinking about it with `buttonPhoto.setEnabled(false);` in the capturePhoto, the onClick of the listener and the capturePhoto are not launched until the `buttonPhoto.setEnabled(true);`, this behaviour is analogues to put that statement in the onClick of the listener, so it's not needed to put it in the listener, it works in both places. This is resolved!
I've written about removing also 91-92 because if the button is disabled the method capturePhoto is not launched actually. Could be `if (!buttonPhoto.isEnabled()) return;` of the line 92 useful now or in the future as a preventive checks or not in your opinion?

EliaBedin replied:
Actually, I forgot to take out the if statement from the old implementation, you are right.
It has the `!buttonPhoto.isEnabled()` because I did a _find and replace all_.
I'm going to take it out now thank you very much.

Commit:
Removed if statement

Removed an if statement that was here from an old implementation.

PrandiniUniPD commented:
The code review is completed, thanks to work with me. Have a nice day.

PrandiniUniPD commented:
Checklist:
- Comments about the code
- Explanations about the method or the code in general
- Modularity of the code
- Readability of the code
- Functionality relativeness to Android tools
- Respect of the Document Style proposed
